cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,681 @@
+  // repeated
+  auto last_edge_with_triangles =
+    thrust::stable_partition(handle.get_thrust_policy(),
+                             edge_triangle_count_pair_first,
+                             edge_triangle_count_pair_first + num_edges,
+                             [] __device__(auto edge_to_num_triangles) {
+                               return thrust::get<1>(edge_to_num_triangles) > 0;
+                             });
+
+  auto last_edge_with_triangles_idx =
+    thrust::distance(edge_triangle_count_pair_first, last_edge_with_triangles);
+  // resize the 'edgelist_srcs' and 'edgelsit_dst'
+  edgelist_srcs.resize(last_edge_with_triangles_idx, handle.get_stream());
+  edgelist_dsts.resize(last_edge_with_triangles_idx, handle.get_stream());
+  num_triangles.resize(last_edge_with_triangles_idx, handle.get_stream());
+}
+
+namespace {
+
+template <typename vertex_t>
+struct exclude_self_loop_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(
+    vertex_t src, vertex_t dst, thrust::nullopt_t, thrust::nullopt_t, thrust::nullopt_t) const
+  {
+    return src != dst
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : thrust::nullopt;
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+struct extract_low_to_high_degree_edges_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(vertex_t src,
+                                                                            vertex_t dst,
+                                                                            edge_t src_out_degree,
+                                                                            edge_t dst_out_degree,
+                                                                            thrust::nullopt_t) const
+  {
+    return (src_out_degree < dst_out_degree)
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : (((src_out_degree == dst_out_degree) &&
+                 (src < dst) /* tie-breaking using vertex ID */)
+                  ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src,
+                                                                                           dst)}
+                  : thrust::nullopt);
+  }
+};
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+struct generate_p_r_q_r {
+  const edge_t edge_first_src_or_dst{};
+  raft::device_span<size_t const> intersection_offsets{};
+  raft::device_span<vertex_t const> intersection_indices{};
+  EdgeIterator transposed_edge_first{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    auto itr = thrust::upper_bound(
+      thrust::seq, intersection_offsets.begin() + 1, intersection_offsets.end(), i);
+    auto idx = thrust::distance(intersection_offsets.begin() + 1, itr);
+
+    if (edge_first_src_or_dst == 0) {
```

Comment:
And if `generate_p_r` is a template non-type parameter, this can be `if constexpr (generate_p_r)`.